File: Puissance4Game/GameLogic/MinimaxAi.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Puissance4Game.GameLogic;

public class MinimaxAi
{
    private readonly int _maxDepth;

    public MinimaxAi(int difficulty)
    {
        _maxDepth = difficulty switch
        {
            <= 1 => 2,
            2 => 4,
            3 => 5,
            >= 4 => 6
        };
    }

    public int ChooseColumn(GameBoard board, int aiPlayer, int humanPlayer)
    {
        var validMoves = board.GetValidMoves().ToList();
        if (!validMoves.Any())
        {
            return -1;
        }

        var orderedMoves = validMoves
            .OrderBy(column => Math.Abs(column - GameBoard.Columns / 2.0))
            .ToList();

        var bestScore = int.MinValue;
        var bestColumn = orderedMoves.First();

        foreach (var column in orderedMoves)
        {
            var row = board.DropPiece(column, aiPlayer);
            var score = Minimax(board, _maxDepth - 1, false, aiPlayer, humanPlayer, int.MinValue, int.MaxValue);
            board.RemovePiece(column, row);

            if (score > bestScore)
            {
                bestScore = score;
                bestColumn = column;
            }
        }

        return bestColumn;
    }

    private int Minimax(GameBoard board, int depth, bool maximizingPlayer, int aiPlayer, int humanPlayer, int alpha, int beta)
    {
        if (depth == 0 || board.IsFull() || board.HasConnectedFour(aiPlayer) || board.HasConnectedFour(humanPlayer))
        {
            return Evaluate(board, aiPlayer, humanPlayer);
        }

        if (maximizingPlayer)
        {
            var value = int.MinValue;
            foreach (var column in board.GetValidMoves())
            {
                var row = board.DropPiece(column, aiPlayer);
                value = Math.Max(value, Minimax(board, depth - 1, false, aiPlayer, humanPlayer, alpha, beta));
                board.RemovePiece(column, row);
                alpha = Math.Max(alpha, value);
                if (alpha >= beta)
                {
                    break;
                }
            }

            return value;
        }
        else
        {
            var value = int.MaxValue;
            foreach (var column in board.GetValidMoves())
            {
                var row = board.DropPiece(column, humanPlayer);
                value = Math.Min(value, Minimax(board, depth - 1, true, aiPlayer, humanPlayer, alpha, beta));
                board.RemovePiece(column, row);
                beta = Math.Min(beta, value);
                if (alpha >= beta)
                {
                    break;
                }
            }

            return value;
        }
    }

    private int Evaluate(GameBoard board, int aiPlayer, int humanPlayer)
    {
        if (board.HasConnectedFour(aiPlayer))
        {
            return 1_000_000;
        }

        if (board.HasConnectedFour(humanPlayer))
        {
            return -1_000_000;
        }

        var score = 0;
        var centerColumn = GameBoard.Columns / 2;
        var centerCount = 0;
        for (var row = 0; row < GameBoard.Rows; row++)
        {
            if (board[row, centerColumn] == aiPlayer)
            {
                centerCount++;
            }
        }

        score += centerCount * 6;

        score += EvaluateLines(board, aiPlayer, humanPlayer);

        return score;
    }

    private int EvaluateLines(GameBoard board, int aiPlayer, int humanPlayer)
    {
        var score = 0;

        // Horizontal
        for (var row = 0; row < GameBoard.Rows; row++)
        {
            var rowValues = new int[GameBoard.Columns];
            for (var column = 0; column < GameBoard.Columns; column++)
            {
                rowValues[column] = board[row, column];
            }

            for (var column = 0; column <= GameBoard.Columns - 4; column++)
            {
                var window = rowValues.Skip(column).Take(4).ToArray();
                score += EvaluateWindow(window, aiPlayer, humanPlayer);
            }
        }

        // Vertical
        for (var column = 0; column < GameBoard.Columns; column++)
        {
            var columnValues = new int[GameBoard.Rows];
            for (var row = 0; row < GameBoard.Rows; row++)
            {
                columnValues[row] = board[row, column];
            }

            for (var row = 0; row <= GameBoard.Rows - 4; row++)
            {
                var window = columnValues.Skip(row).Take(4).ToArray();
                score += EvaluateWindow(window, aiPlayer, humanPlayer);
            }
        }

        // Positive slope diagonals
        for (var row = 0; row <= GameBoard.Rows - 4; row++)
        {
            for (var column = 0; column <= GameBoard.Columns - 4; column++)
            {
                var window = new[]
                {
                    board[row, column],
                    board[row + 1, column + 1],
                    board[row + 2, column + 2],
                    board[row + 3, column + 3]
                };
                score += EvaluateWindow(window, aiPlayer, humanPlayer);
            }
        }

        // Negative slope diagonals
        for (var row = 3; row < GameBoard.Rows; row++)
        {
            for (var column = 0; column <= GameBoard.Columns - 4; column++)
            {
                var window = new[]
                {
                    board[row, column],
                    board[row - 1, column + 1],
                    board[row - 2, column + 2],
                    board[row - 3, column + 3]
                };
                score += EvaluateWindow(window, aiPlayer, humanPlayer);
            }
        }

        return score;
    }

    private int EvaluateWindow(IReadOnlyList<int> window, int aiPlayer, int humanPlayer)
    {
        var aiCount = window.Count(value => value == aiPlayer);
        var humanCount = window.Count(value => value == humanPlayer);
        var emptyCount = window.Count(value => value == 0);

        var score = 0;

        if (aiCount == 4)
        {
            score += 5_000;
        }
        else if (aiCount == 3 && emptyCount == 1)
        {
            score += 150;
        }
        else if (aiCount == 2 && emptyCount == 2)
        {
            score += 20;
        }

        if (humanCount == 4)
        {
            score -= 5_000;
        }
        else if (humanCount == 3 && emptyCount == 1)
        {
            score -= 180;
        }
        else if (humanCount == 2 && emptyCount == 2)
        {
            score -= 18;
        }

        return score;
    }
}

File: Puissance4Android/app/src/main/java/com/example/puissance4/MinimaxAi.java
package com.example.puissance4;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class MinimaxAi {
    private final int maxDepth;

    public MinimaxAi(int difficulty) {
        if (difficulty <= 1) {
            maxDepth = 2;
        } else if (difficulty == 2) {
            maxDepth = 4;
        } else if (difficulty == 3) {
            maxDepth = 5;
        } else {
            maxDepth = 6;
        }
    }

    public int chooseColumn(GameBoard board, int aiPlayer, int humanPlayer) {
        List<Integer> validMoves = board.getValidMoves();
        if (validMoves.isEmpty()) {
            return -1;
        }

        List<Integer> orderedMoves = new ArrayList<>(validMoves);
        double center = GameBoard.COLUMNS / 2.0;
        Collections.sort(orderedMoves, Comparator.comparingDouble(column -> Math.abs(column - center)));

        int bestScore = Integer.MIN_VALUE;
        int bestColumn = orderedMoves.get(0);

        for (int column : orderedMoves) {
            int row = board.dropPiece(column, aiPlayer);
            int score = minimax(board, maxDepth - 1, false, aiPlayer, humanPlayer, Integer.MIN_VALUE, Integer.MAX_VALUE);
            board.removePiece(column, row);

            if (score > bestScore) {
                bestScore = score;
                bestColumn = column;
            }
        }

        return bestColumn;
    }

    private int minimax(GameBoard board, int depth, boolean maximizingPlayer, int aiPlayer, int humanPlayer, int alpha, int beta) {
        if (depth == 0 || board.isFull() || board.hasConnectedFour(aiPlayer) || board.hasConnectedFour(humanPlayer)) {
            return evaluate(board, aiPlayer, humanPlayer);
        }

        if (maximizingPlayer) {
            int value = Integer.MIN_VALUE;
            for (int column : board.getValidMoves()) {
                int row = board.dropPiece(column, aiPlayer);
                value = Math.max(value, minimax(board, depth - 1, false, aiPlayer, humanPlayer, alpha, beta));
                board.removePiece(column, row);
                alpha = Math.max(alpha, value);
                if (alpha >= beta) {
                    break;
                }
            }
            return value;
        }

        int minimizingValue = Integer.MAX_VALUE;
        for (int column : board.getValidMoves()) {
            int row = board.dropPiece(column, humanPlayer);
            minimizingValue = Math.min(minimizingValue, minimax(board, depth - 1, true, aiPlayer, humanPlayer, alpha, beta));
            board.removePiece(column, row);
            beta = Math.min(beta, minimizingValue);
            if (alpha >= beta) {
                break;
            }
        }
        return minimizingValue;
    }

    private int evaluate(GameBoard board, int aiPlayer, int humanPlayer) {
        if (board.hasConnectedFour(aiPlayer)) {
            return 1_000_000;
        }
        if (board.hasConnectedFour(humanPlayer)) {
            return -1_000_000;
        }

        int score = 0;
        int centerColumn = GameBoard.COLUMNS / 2;
        int centerCount = 0;
        for (int row = 0; row < GameBoard.ROWS; row++) {
            if (board.getCell(row, centerColumn) == aiPlayer) {
                centerCount++;
            }
        }
        score += centerCount * 6;
        score += evaluateLines(board, aiPlayer, humanPlayer);
        return score;
    }

    private int evaluateLines(GameBoard board, int aiPlayer, int humanPlayer) {
        int score = 0;

        // Horizontal
        for (int row = 0; row < GameBoard.ROWS; row++) {
            for (int column = 0; column <= GameBoard.COLUMNS - 4; column++) {
                int[] window = new int[4];
                for (int i = 0; i < 4; i++) {
                    window[i] = board.getCell(row, column + i);
                }
                score += evaluateWindow(window, aiPlayer, humanPlayer);
            }
        }

        // Vertical
        for (int column = 0; column < GameBoard.COLUMNS; column++) {
            for (int row = 0; row <= GameBoard.ROWS - 4; row++) {
                int[] window = new int[4];
                for (int i = 0; i < 4; i++) {
                    window[i] = board.getCell(row + i, column);
                }
                score += evaluateWindow(window, aiPlayer, humanPlayer);
            }
        }

        // Positive slope diagonals
        for (int row = 0; row <= GameBoard.ROWS - 4; row++) {
            for (int column = 0; column <= GameBoard.COLUMNS - 4; column++) {
                int[] window = new int[4];
                for (int i = 0; i < 4; i++) {
                    window[i] = board.getCell(row + i, column + i);
                }
                score += evaluateWindow(window, aiPlayer, humanPlayer);
            }
        }

        // Negative slope diagonals
        for (int row = 3; row < GameBoard.ROWS; row++) {
            for (int column = 0; column <= GameBoard.COLUMNS - 4; column++) {
                int[] window = new int[4];
                for (int i = 0; i < 4; i++) {
                    window[i] = board.getCell(row - i, column + i);
                }
                score += evaluateWindow(window, aiPlayer, humanPlayer);
            }
        }

        return score;
    }

    private int evaluateWindow(int[] window, int aiPlayer, int humanPlayer) {
        int aiCount = 0;
        int humanCount = 0;
        int emptyCount = 0;

        for (int value : window) {
            if (value == aiPlayer) {
                aiCount++;
            } else if (value == humanPlayer) {
                humanCount++;
            } else {
                emptyCount++;
            }
        }

        int score = 0;

        if (aiCount == 4) {
            score += 5_000;
        } else if (aiCount == 3 && emptyCount == 1) {
            score += 150;
        } else if (aiCount == 2 && emptyCount == 2) {
            score += 20;
        }

        if (humanCount == 4) {
            score -= 5_000;
        } else if (humanCount == 3 && emptyCount == 1) {
            score -= 180;
        } else if (humanCount == 2 && emptyCount == 2) {
            score -= 18;
        }

        return score;
    }
}
